cohort-evaluator-spark/src/main/java/com/ibm/cohort/cql/terminology/R4FileSystemFhirTerminologyProvider.java

```diff
@@ -0,0 +1,227 @@
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.PathFilter;
+import org.cqframework.cql.elm.execution.VersionedIdentifier;
+import org.hl7.fhir.r4.model.ValueSet;
+import org.hl7.fhir.r4.model.ValueSet.ConceptReferenceComponent;
+import org.hl7.fhir.r4.model.ValueSet.ConceptSetComponent;
+import org.opencds.cqf.cql.engine.runtime.Code;
+import org.opencds.cqf.cql.engine.terminology.CodeSystemInfo;
+import org.opencds.cqf.cql.engine.terminology.ValueSetInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ca.uhn.fhir.context.ConfigurationException;
+import ca.uhn.fhir.context.FhirContext;
+import ca.uhn.fhir.parser.DataFormatException;
+
+/**
+ * This class is used to read ValueSet definitions from the local filesystem or S3 compatible source.
+ * ValueSet definitions are expected to be stored in FHIR xml or JSON format and can be generated from
+ * spreadsheets using the com.ibm.cohort.tooling.fhir.ValueSetImporter tool.
+ * 
+ * In order to use ValueSets stored on the filesystem, the following spark config parameters must be provided:
+ * -Dspark.hadoop.fs.s3a.access.key = access_key_value
+ * -Dspark.hadoop.fs.s3a.secret.key = secret_key_value
+ * -Dspark.hadoop.fs.s3a.endpoint = object store endpoint
+ * 
+ * as well as setting the following environment variables:
+ * AWS_ACCESS_KEY = access_key_value
+ * AWS_ENDPOINT = object store endpoint
+ * AWS_LOCATION = location value ie us-east
+ * AWS_SECRET_ACCESS_KEY = secret_key_value
+ *
+ */
+public class R4FileSystemFhirTerminologyProvider implements CqlTerminologyProvider {
+	private Path terminologyDirectory;
+	private Configuration configuration;
+	
+	private static final Logger LOG = LoggerFactory.getLogger(R4FileSystemFhirTerminologyProvider.class);
+	
+	private static final FhirContext fhirContext = FhirContext.forR4();
+	
+	private Map<VersionedIdentifier, ValueSet> valueSetCache = new HashMap<>();
+	
+	public R4FileSystemFhirTerminologyProvider(Path terminologyDirectory, Configuration configuration) {
+		super();
+		this.terminologyDirectory = terminologyDirectory;
+		LOG.info("TerminologyDirectory is " + terminologyDirectory.toString());
+		this.configuration = configuration;
+	}
+
+	/* (non-Javadoc)
+	 * 
+	 * Returns true if the provided code exists in the provided ValueSet
+	 * 
+	 * @see org.opencds.cqf.cql.engine.terminology.TerminologyProvider#in(org.opencds.cqf.cql.engine.runtime.Code, org.opencds.cqf.cql.engine.terminology.ValueSetInfo)
+	 */
+	@Override
+	public boolean in(Code code, ValueSetInfo valueSetInfo) {
+		LOG.debug("Entry: in() ValueSet.getId=[" + valueSetInfo.getId() + "] version=[" + valueSetInfo.getVersion() +"]");
```

Comment:
SLF4J provides parameterized message support out of the box. You can use Java MessageFormat style syntax (e.g. {} placeholders) and then pass the arguments for those in a varargs parameter after the message. It allows for SLF4J to short circuit string creation when the logging is disabled. This is similar to why people use log guards (e.g. if( LOG.isDebugEnabled()) in some older log4j code. It is recommended to use the parameterized or guarded version for optimal performance. Instead of me tagging every line with this note, please consider this comment applicable for all log messages that come after.